Fix HealthTest::testBaseUrlHasBeenSet()

kenjis opened the pull request:
**Description**
Fixes #3977

Tests about `baseURL`:
- `baseURL` is a valid URL
- Set it in `.env` or `app/Config/App.php`

**Checklist:**
- [x] Securely signed commits
- [ ] Component(s) with PHPdocs
- [ ] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

kenjis commented:
```sh-session
$ composer create-project codeigniter4/appstarter ci4
$ cd ci4/
$ vendor/bin/phpunit --no-coverage
PHPUnit 9.5.2 by Sebastian Bergmann and contributors.

....F                                                               5 / 5 (100%)

Time: 00:00.085, Memory: 12.00 MB

There was 1 failure:

1) HealthTest::testBaseUrlHasBeenSet
baseURL in app/Config/App.php is not set
Failed asserting that false is true.

/Users/kenji/tmp/ci4/tests/unit/HealthTest.php:38

FAILURES!
Tests: 5, Assertions: 7, Failures: 1.
```

kenjis commented:
```diff
--- a/app/Config/App.php
+++ b/app/Config/App.php
@@ -23,7 +23,7 @@ class App extends BaseConfig
     *
     * @var string
     */
-   public $baseURL = 'http://localhost:8080/';
+   public $baseURL = 'http://abc.example.com/';
 
    /**
     * --------------------------------------------------------------------------

```

```sh-session
$ vendor/bin/phpunit --no-coverage
PHPUnit 9.5.2 by Sebastian Bergmann and contributors.

.....                                                               5 / 5 (100%)

Time: 00:00.080, Memory: 12.00 MB

OK (5 tests, 8 assertions)
```

kenjis commented:
```sh-session
$ git reset --hard HEAD
$ cp env .env
$ vi .env
```

```diff
--- env	2021-02-01 11:56:38.000000000 +0900
+++ .env	2021-02-07 11:09:59.000000000 +0900
@@ -20,7 +20,7 @@
 # APP
 #--------------------------------------------------------------------
 
-# app.baseURL = ''
+app.baseURL = http://xyz.example.com/
 # app.forceGlobalSecureRequests = false
 
 # app.sessionDriver = 'CodeIgniter\Session\Handlers\FileHandler'
```

```sh-session
$ vendor/bin/phpunit --no-coverage
PHPUnit 9.5.2 by Sebastian Bergmann and contributors.

.....                                                               5 / 5 (100%)

Time: 00:00.078, Memory: 12.00 MB

OK (5 tests, 8 assertions)
```

kenjis commented:
I think this is probably OK. But a bit complicated...

MGatner commented:
Per the issue conversation this should be updated only to fix the regex. But please make sure the regex allows single quotes still, the version in this PR does not.

kenjis commented:
URL with single quotes passes the test. `'` is included in `.`.

```diff
--- env	2021-02-01 11:56:38.000000000 +0900
+++ .env	2021-02-10 10:57:00.000000000 +0900
@@ -20,7 +20,7 @@
 # APP
 #--------------------------------------------------------------------
 
-# app.baseURL = ''
+app.baseURL = 'http://localhost:8080/'
 # app.forceGlobalSecureRequests = false
 
 # app.sessionDriver = 'CodeIgniter\Session\Handlers\FileHandler'
```

```sh-session
$ vendor/bin/phpunit --no-coverage
PHPUnit 9.5.2 by Sebastian Bergmann and contributors.

.....                                                               5 / 5 (100%)

Time: 00:00.075, Memory: 12.00 MB

OK (5 tests, 8 assertions)
```

paulbalandan commented:
why are we disallowing `localhost:8080` for the baseURL?

kenjis commented:
The user guide says:
> Open the app/Config/App.php file with a text editor and set your base URL. If you need more flexibility, the baseURL may be set within the .env file as app.baseURL=”http://example.com/”. (Always use a trailing slash on your base URL!)
https://codeigniter4.github.io/CodeIgniter4/installation/running.html#initial-configuration-set-up

So I think users are supposed to set their base URL, and `localhost:8080` is the default, they did not set it.

MGatner commented:
Localhost is fine for a base URL! See my comment above, but the issue is when a user has explicitly created a scenario where the base URL is blank - out of the box it should work fine.

MGatner commented:
I think the regex currently fails to catch `app.baseUrl = ''` which should fail.

kenjis commented:
How about this? I've cleaned up.

I removed disallowing `localhost:8080`, and changed the validation to use Validator.

`app.baseURL = ''` is not a valid URL, so it fails.

paulbalandan commented on admin/starter/tests/unit/HealthTest.php:
```diff
@@ -16,18 +18,35 @@ public function testIsDefinedAppPath()
 
-		// Then check the actual config file
+		// Get the baseURL in app/Config/App.php
+		// Don't forget phpunit.xml.dist sets app.baseURL
```
I think this is not needed. As `ConfigReader` does not call `Config\App`'s constructor, its properties would not be populated by environment values but only by the hardcoded values.

kenjis replied:
@paulbalandan I fixed the comment. My intention is why we need to use ConfigReader.
I forgot about phpunit.xml.dist at the first time.

bereldhuin commented:
I noticed another problem with the regexp (using CI 4.1.1) : VisualStudio code tends to remove spaces in .env files.

So my baseUrl looks like this : 
`app.baseURL='http://192.168.1.11:81/'`
Instead of
`app.baseURL = 'http://192.168.1.11:81/'`

Which makes the phpunit fail on testBaseUrlHasBeenSet
